### PR TITLE
Bump F# to 4.5

### PR DESCRIPTION
**Pull request by nosami**
make pkg

### CODE REVIEW

File: packaging/MacSDK/fsharp.py [hunk 1]
```diff
@@ -2,15 +2,13 @@ class FsharpPackage(GitHubTarballPackage):
...
+            '4.5.0',
+            '3de387432de8d11a89f99d1af87aa9ce194fe21b',
             configure='./configure --prefix="%{package_prefix}"',
             override_properties={ 'make': 'make' })
 
         self.extra_stage_files = ['lib/mono/xbuild/Microsoft/VisualStudio/v/FSharp/Microsoft.FSharp.Targets']
         self.sources.extend(['patches/fsharp-portable-pdb.patch',
-                             'patches/fsharp-string-switchName.patch',
-                             'patches/fsharp-path-overloads.patch',
                              'patches/fsharp-debug-pinvoke-fix.patch'])
```
Comment (akoeplinger) [inline packaging/MacSDK/fsharp.py]: @nosami btw do you think we can remove those two remaining patches as well now?

Comment (nosami) [reply packaging/MacSDK/fsharp.py]: No. Not yet. I will try and get them upstreamed.

**Comment by akoeplinger** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by akoeplinger** [comment]
@dsyme you don't seem to be in the Microsoft/Xamarin/Mono GitHub orgs (at least not publicly) so the GitHub trigger won't work for you.

The failures from the previous build https://jenkins.mono-project.com/job/build-package-osx-mono-pullrequest/246 should still be accurate though since there wasn't another F# bump.

**Comment by nosami** [comment]
@akoeplinger The errors now seem to be an artifact of previous test runs. The error that we now see is

```
15:49:58 mkdir: @/external/bockbuild/scratch/fsharp.install@/external/bockbuild/stage/lib/mono/fsharp: Permission denied
15:49:58 
make[1]: *** [install-sdk-lib] Error 1
15:49:58 make: *** [install] Error 2
15:49:58 
15:49:58 fsharp->build_sh:
```

As far as I can tell, the permission error has been caused by an inadvertent `sudo make install` from a previous build. I think the bots need manual clean up.

**Comment by akoeplinger** [comment]
Ok, I cleaned the bots, let's try another build.

@monojenkins build pkg

**Comment by akoeplinger** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
Same error after clean up. I have no idea where the 'Permission denied' is coming from.

**Comment by nosami** [comment]
I am assuming that there is some issue with the Jenkins bots somewhere. The same script builds on Travis. 

I don't see `sudo` being used anywhere apart from `sudo make install` (which I patched out - It appears that the bots run `make install` to a custom folder). It's also used for restoring packages when the script is ran on linux.

**Comment by dsyme** [comment]
Who can add me to those orgs? I should be in xamarin I thought

Can we get direct access to one of these build machines?

**Comment by akoeplinger** [comment]
@dsyme yeah you are in the org but your membership is set to private (so the bot can't see it). Go to https://github.com/orgs/xamarin/people?utf8=%E2%9C%93&query=dsyme and flip it to "public"

**Comment by akoeplinger** [comment]
@nosami yes, let's talk on Slack

**Comment by alexischr** [comment]
It seemed like more uncleanable files from `sudo make install`

**Comment by alexischr** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by nosami** [comment]
@monojenkins build pkg

**Comment by dsyme** [comment]
Notes to self....

Prior to this PR we did this:
```
        self.sh('autoreconf') 
        Package.configure(self)
        Package.make(self)
```
I'm told that the last two lines are just `configure && make` though we should double check this.

After this PR we do this:
```
        self.sh('./mono/build-mono.sh')
```
Which contains this:
```
make && make install
```
after patching. Prior to patching this was:
```
 ./mono/prepare-mono.sh && \
make && \
sudo make install
```

**Comment by dsyme** [comment]
I have created a no-op PR so we can get a bockbuild baseline for this to understand why the change to the F# build process is problematic https://github.com/mono/mono/pull/12018

**Comment by dsyme** [comment]
@monojenkins build pkg

**Comment by dsyme** [comment]
@monojenkins build pkg

**Comment by dsyme** [comment]
@monojenkins build pkg

**Comment by dsyme** [comment]
@monojenkins build pkg

**Comment by dsyme** [comment]
@monojenkins build pkg

**Comment by dsyme** [comment]
@directhex 

Quick question if possible…  We need to install into the freshly built mono package directories, like we’ve always done.  In particular we used to have this
	 
            configure='./configure --prefix="%{package_prefix}"',	 
		
which must have worked somehow, but now have this

            override_properties={ 'make': 'make all install prefix="%{package_prefix}"' })	 
		
which is intended to do the same thing, but strangely gives this:

        command failed: make all install prefix="/Library/Frameworks/Mono.framework/Versions/5.23.0"

which as you point out is wrong wrong wrong. I _thought_ this would give an install into `/external/bockbuild/stage/lib/mono`.  Perhaps `package_prefix` isn’t what we think it is or the old configure was ignoring it. I would have expected `package_prefix` to be the location of the built Mono bits…

Just to check – do you know what we expect `package_prefix` to be here?  Is there a way to know where the mono bits are installed and to make sure e’re running with them?

I’ll kick of a build without package_prefix now.  The default mechanism we use is this:

    monocmd = $(shell which mono)
    monocmddir = $(dir $(monocmd))
    prefix = $(shell (cd $(monocmddir)/..; pwd))

**Comment by dsyme** [comment]
Quick update: 

Without “prefix=…” the automatically inferred paths look correct. However it looks like our “make install” is not able to “mkdir” where we need to:

	/external/bockbuild/stage/lib/mono/fsharp

I’m not understanding why this was working previously.  I’ll try to revert everything and get a log.

**Comment by dsyme** [comment]
@monojenkins build pkg

**Comment by dsyme** [comment]
@nosami OK, I did a force push - first commit contains everything up to the point where we were trying `prefix=...` and second commit does a revert of that commit back to master. 

Let's get a log of this and work out WTF is going on

**Comment by marek-safar** [comment]
@monojenkins build

**Comment by auduchinok** [comment]
Any update on this PR?

**Comment by dsyme** [comment]
We're really struggling to de ug the build of the package via CI, it's a nightmare. Nothing makes sense in the logs.

We do have to get back to this

**Comment by directhex** [comment]
@monojenkins build pkg

**Comment by marek-safar** [comment]
This PR is no-op, could you revert the revert or open a new one?

**Comment by directhex** [comment]
Yeah, @dsyme @nosami if you can give this another shot, I'll work with you on making it happen. High-importance task.

**Comment by dsyme** [comment]
Oddly "OS X Package — Build finished." is now reported green.  The three "red" aren't anything to do with the F# update I'd think, though we could check...

@directhex Is there any way I could get a log on to a machine that is able to build the packages end to end?

**Comment by directhex** [comment]
That's down to RE, who administers those machines. @cadsit, @MSylvia? https://jenkins.mono-project.com/label/mono-package/ is the label which does package builds

**Comment by dsyme** [comment]
I reinstated the changes in this PR - I'd forgotten that we'd reverted them in order to get clean log files

I'm certain we can get this through easily if we have access to a machine doing the CI builds or capable of doing equivalent package builds

**Comment by directhex** [comment]
@monojenkins build pkg

**Comment by directhex** [comment]
Why it failed:

This patch doesn't apply: https://github.com/mono/mono/blob/88744920e71fc6d9506710a8c3079834e5110058/packaging/MacSDK/patches/fsharp-msbuild-16-0.patch

There's no file `mono/config.make.in` in the fsharp commit pointed to by fsharp.py: https://github.com/fsharp/fsharp/tree/3de387432de8d11a89f99d1af87aa9ce194fe21b/mono

**Comment by dsyme** [comment]
@monojenkins build pkg

**Comment by dsyme** [comment]
@directhex I do feel like we'll need access to a build machine to make progress on this.  Hand editing patches on an 18h turnaround cycle is not fun ....

**Comment by dsyme** [comment]
@monojenkins build pkg

**Comment by directhex** [comment]
@dsyme I've spent some time on the machine which built your last package, and can roughly summarize:

1. Bockbuild expects packages to install into a scratch directory `@/external/bockbuild/scratch/packagename.install`
1. Bockbuild then copies files from there into the "final" staging directory, `@/external/bockbuild/stage/`, _with permissions set read-only since the final package needs to be read-only_
1. FSharp is failing with "permission denied" because it is automatically detecting the Mono directory in `/stage/` and trying to make changes in that "final" prefix

I'm experimenting with fixes.

**Comment by dsyme** [comment]
@directhex any luck?

**Comment by directhex** [comment]
Sorry, I was out yesterday. Back on it today.

**Comment by directhex** [comment]
@dsyme so i made some tweaks so far, and made some progress - by removing https://github.com/mono/mono/blob/94a00c9b797047d9f3dc624945ce7dcd81777dbb/packaging/MacSDK/fsharp.py#L8 that forces bockbuild to run the `all` and `install` steps separately, and bockbuild defines DESTDIR correctly when calling `install` (but doesn't for `all`, which is why DESTDIR wasn't being set previously).

This exposes a new issue, however, because the `install-sdk-lib` target (and, really, anything under `install`) appends `monolibdir` to `DESTDIR`, and those are _both_ absolute, not relative. So current result is:

`mkdir: @/external/bockbuild/scratch/fsharp.install@/external/bockbuild/stage/lib/mono/fsharp: Permission denied`

Which is wrong and silly.

Working on a cure.

**Comment by dsyme** [comment]
@directhex Thank you! Both for the update and the hard work :)

**Comment by dsyme** [comment]
@directhex Any progress on this?

**Comment by directhex** [comment]
I drafted in @alexischr to help, and he's able to reproduce the issue but not yet work out why it's happening.

**Comment by alexischr** [comment]
We have just figured out the issue - bockbuild expects `make install` to have the installation end up in `$(DESTDIR)/$(prefix)` where `DESTDIR` here is a local scratch directory while `prefix` is the final destination path (e.g. `/Library/Frameworks/Mono.Framework/...`). The patch and the bockbuild manifest will be adjusted for that.

**Comment by alexischr** [comment]
Actually a much better explanation is this: The following line is problematic in the scope of building the Mac SDK package (using bockbuild):

https://github.com/fsharp/fsharp/blob/ccfccbacfdb16b2ab143cf83f5ece20f738fd864/mono/config.make#L5

It's problematic because where `mono` lives in this case would be something like `/macsdk-staging-dir/bin/mono` which is not the actual prefix (prefix here being the final destination, `/Library/Frameworks/Mono.framework/...`) but a staging area that bockbuild uses (this is how we build on system prefixes without installing on CI)

Bockbuild expects things to end up  in `/fsharp-staging-dir/$(prefix)` (because that's what most `make install`s do) and if things end up in another directory in `/fsharp-staging-dir/`  (like `/fsharp-staging-dir/macsdk-staging-dir` in our case) that would cause us to miss those files (this has happened before, with packages that have broken use of `DESTDIR`). So there is an added write-protection for anywhere in `/fsharp-staging-dir` that *isn't*  `/fsharp-staging-dir/$(prefix)` , because a write-protection error, while obtuse, is much preferable to missing the fact that some files are not packaged (as would have happened here)

So the fix is to stop trying to infer the prefix, and return to providing it explicitly to the build system (this used to be the case via the `configure` step)

**Comment by alexischr** [comment]
@monojenkins build pkg

**Comment by alexischr** [comment]
@monojenkins build pkg

**Comment by alexischr** [comment]
@monojenkins build pkg

**Comment by dsyme** [comment]
@alexischr Thank you so much for looking into this - looks as though you're close to solving the problem?

**Comment by directhex** [comment]
Looks like progress. Gotta use `sn` from PATH not PREFIX though

**Comment by alexischr** [comment]
@monojenkins build pkg

**Comment by alexischr** [comment]
@directhex given our understanding of the problem now, do you think it's still necessary that both `all` and `install` are run on the `build` step? I may fix that up in a subsequent PR.

**Comment by directhex** [comment]
@alexischr my strong preference is that they NOT be run on the same build step, tbh

**Comment by alexischr** [comment]
It built!! @directhex sounds good, I am not even going to try here though. Merging!

**Comment by dsyme** [comment]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 

🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾 🍾

**Comment by dsyme** [comment]
Thank you so, so much to everyone who helped with this little piece of nightmare.

**Comment by auduchinok** [comment]
BTW, should we talk about bumping it to 4.6 already? :)

**Comment by dsyme** [comment]
@auduchinok First we drink all that champagne I just posted

**Comment by alexischr** [comment]
@monojenkins backport to 2019-02

**Comment by alexischr** [comment]
@monojenkins backport to 2018-10

**Comment by alexischr** [comment]
I'll do the backporting manually. And yeah we should get started on 4.6 right away :)

**Comment by auduchinok** [comment]
@alexischr Thank you again!
And just to get some insight on Mono release process: what Mono versions this change should be expected to get to and when to expect that to happen?

**Comment by alexischr** [comment]
@monojenkins backport to 2019-04

**Comment by alexischr** [comment]
@auduchinok I am attempting backports from master to 6.0 and 6.2 (see PRs referring to this one). Otherwise this will make the next bi-monthly branch (6.4).

These are unreleased/unstable Mono versions and have not finished being integrated with Xamarin products, but 6.0 is on preview as a standalone Mac package: https://www.mono-project.com/download/preview/ . If this merges on 6.0, we can update the 6.0 preview to include it. Do note - this package is not guaranteed or even expected to work with any releases of Visual Studio for Mac.

As we were not targeting this for a specific Mono/Xamarin release (and this is a big change), I did not end up backporting to stable branches.

**Comment by auduchinok** [comment]
Thanks. I was naively expecting we could get it in a 5.x release. 🙂

**Comment by directhex** [comment]
RPMs are done. Debs are blocked by https://github.com/mono/mono/issues/14103

**Pull request merged**